fix issue 17243 - std.math.{FloatingPointControl,ieeeFlags} don't work on x86_64

aG0aep6G opened the pull request:
I'm far from being an expert on this stuff. I learned about the different control registers only when investigating this issue here. If someone more experienced wants to take over, be my guest.

I have a [test file for unmasked exceptions](https://gist.github.com/aG0aep6G/bcf65e37cc4a9e5f71a96b4c20bbc3e7), but I don't know how to incorporate it into the phobos tests.

A test for `roundToNearest` is missing, because I'm having a hard time thinking of one.

aG0aep6G commented:
It's my understanding that the issue can also manifest on 32-bit x86s when the compiler is configured to emit SSE instructions (not sure if that can be done with dmd).There doesn't seem to be a `version` for SSE, so I don't know how to approach that variant of the issue.

----

The failures of the doc tester and CircleCI don't make sense to me. The doc tester says:

>../phobos/std/traits.d(6620): Error: static assert  "Type uint does not have an Unsigned counterpart"

And CircleCI says:

> out/std_regex_package.d(43): Error: undefined identifier 'popFrontN'

Both seem unrelated.

BBasile commented:
> It's my understanding that the issue can also manifest on 32-bit x86s when the compiler is configured to emit SSE instructions (not sure if that can be done with dmd)

On x86, even if for rounding DMD emitts code for the FPU one may use the std.math API and write custom asm SSE code that relies on it, hence both FPU and SSE flags should be set. (IMO, this can be discussed)

aG0aep6G commented:
> On x86, even if for rounding DMD emitts code for the FPU one may use the std.math API and write custom asm SSE code that relies on it, hence both FPU and SSE flags should be set.

I've added a fallback to CPUID for x86.

----

From my end, this is good to go. I've fixed the test failures and squashed the fixup commits. auto-tester is green (pending testing of the fallback mentioned above).

aG0aep6G commented on std/math.d:
```diff
@@ -155,6 +155,18 @@ else version(D_InlineAsm_X86_64)
 }
 
+version (X86_64) version = StaticallyHaveSSE;
+version (X86) version (OSX) version = StaticallyHaveSSE;
```
I'm not sure if this is correct. Does OSX require SSE? This is here, because I've noticed that dmd emits SSE code for 32-bit OSX.

MartinNowak replied:
It's b/c Apple only ever sold x86_64 CPUs.

aG0aep6G commented on std/math.d:
```diff
@@ -4406,10 +4418,9 @@ private:
             INVALID_MASK   = 0x01
         }
-        // Don't bother about subnormals, they are not supported on most CPUs.
-        //  SUBNORMAL_MASK = 0x02;
```
Comment seems bogus. My run-of-the-mill AMD CPU and the auto-tester ones apparently support it.

aG0aep6G commented on std/math.d:
```diff
@@ -4531,6 +4538,10 @@ public:
 
+     /** A subnormal number was used in a calculation.
+     (example: x = real.min_normal / 2 + 1) */
+     @property bool subnormal() { return (flags & SUBNORMAL_MASK) != 0; }
```
@andralex: Adding a new method here. As far as I remember, you want to review new names.

andralex replied:
I'm cool, please make `const` and so the others

aG0aep6G replied:
> please make const and so the others

Made this one const. New PR for the others as it's independent from this PR: https://github.com/dlang/phobos/pull/5246

aG0aep6G commented on std/math.d:
```diff
@@ -4893,24 +4964,33 @@ private:
-                    fclex;
-                    fldcw   8[RSP];
-                    ret;
-                }
```
This deleted code looks very clever. Unfortunately, there is no comment on what it achieves. The generic variant below seems to work just fine on Win64, and it's much simpler. So I'm tearing down this fence without knowing why it was put up.

MartinNowak replied:
It was just the naked version of the asm block below, no need for that here.

aG0aep6G commented:
> auto-tester is green (pending testing of the fallback mentioned above).

Of course it went red then. Fixed getIeeeFlags and resetIeeeFlags. All green again.

ibuclaw commented on std/math.d:
```diff
@@ -4531,6 +4545,10 @@ public:
 
+     /** A subnormal number was used in a calculation.
+     (example: x = real.min_normal / 2 + 1) */
+     @property bool subnormal() const { return (flags & SUBNORMAL_MASK) != 0; }
```
The majority of targets don't have an exception for denormals, I don't think this should be added here.

aG0aep6G replied:
> The majority of targets don't have an exception for denormals, I don't think this should be added here.

As far as I see, IeeeFlags only works on x86 and x86-64 at the moment. Everything else errors out with "not yet supported". x86(-64) are the ones that have this, right? Would it be ok if this was `version`ed away from the others?

jpf91 replied:
> As far as I see, IeeeFlags only works on x86 and x86-64 at the moment. Everything else errors out with "not yet supported"

GDC has code for ARM as well: https://github.com/D-Programming-GDC/GDC/blob/master/libphobos/src/std/math.d#L4372 and LDC probably supports even more targets. The part of the code that has never been upstreamed is mostly GDC extended ASM so it's not useful for other compilers anyway.

We unfortunately don't have any guidelines yet how to deal with architecture specific features (there's a similar problem for `cpuid`). I guess we could add this function but we should at least make sure to explicitly document it as non-portable.

aG0aep6G commented:
I've removed the new `subnormal` method as it's not really related to issue 17243. The method can be discussed and added separately.

I've also squashed everything into one commit. The commit list was becoming messy.
